(dev/core#4433) - Implement Civi::url() with prefixes and OOP enhancements

totten opened the pull request:
Overview
----------------------------------------

This is a concrete demonstration of some ideas from https://lab.civicrm.org/dev/core/-/issues/4433.

Before
----------------------------------------

This function exists:

```php
CRM_Utils_System::url(NULL, NULL, FALSE, 1, '', TRUE, __CLASS__, NULL, TRUE);
```

And that does... something. Have fun figuring it out...

After
----------------------------------------

```php
echo Civi::url('frontend://civicrm/event/info?id=123&reset=1');

echo Civi::url('//civicrm/profile/view?reset=1')
  ->addQuery(['gid' => $x, 'id' => $y])

echo Civi::url('assetBuilder://crm-l10n.js')
  ->addQuery(['locale' => $x])
```

(*Docblock has more examples...*)

Technical Details
----------------------------------------

In this rendition:

* The entry-point (frontend/backend/etc) is a URI scheme (`frontend://`).
* It respects URI convention for inheriting current scheme (`//foo.txt` means "get `foo.txt` with current scheme)
* The list of URI schemes is borrowed from other places in core -- e.g. the Guzzle-test system has `frontend://ROUTE` and `backend://ROUTE`. Some of the asset/resource management supports `assetBuilder://NAME` and `ext://EXT/FILE`.

totten committed:
CRM_Core_Menu - Add helper method, `isPublicRoute($path)`.

totten committed:
Mark civicrm/asset/builder as public

totten committed:
Implement Civi::url()

colemanw commented on Civi/Core/Url.php:
```diff
@@ -0,0 +1,353 @@
+   *
+   * @var bool
+   */
+  private $htmlEscape = FALSE;
```
LOL my first thought when I saw the title of this PR was "please can we kill off `$htmlize`"!
Guess not... but at least you called it insane, so if not the death penalty, at least it received a good taunting.

colemanw commented on Civi/Core/Url.php:
```diff
@@ -0,0 +1,353 @@
+
+        // (p)lain text encoding
+        case 'p':
+          $this->htmlEscape = FALSE;
```
Ooh, an undocumented flag that overrules `htmlize`! Yet another blow to its fragile ego... perhaps leading it to the brink of a well-deserved existential crisis.

colemanw replied:
@totten this flag is still undocumented. Should it be removed?

totten replied:
The "undocumented" note had confused me, but I see now: it's documented on `Civi::url(...$flags...)` but not `Url::useFlags($flags)`. That's easy to cleanup.

I should take a stab at the Smarty side before deciding whether to remove that flag. My suspicion is that it'll make it easier to tune Smarty. (Quick estimate: in core, there are ~530 places where Smarty wants HTML and ~130 places where it wants plain text.)

colemanw commented on Civi/Core/Url.php:
```diff
@@ -0,0 +1,353 @@
+      $result = 'http:' . substr($result, 6);
+    }
+
+    return $this->htmlEscape ? htmlentities($result) : $result;
```
But seriously, can we just not include this oddball feature in the initial implementation and hold off on adding it unless some pressing reason emerged in the future. I'd much rather have a new function be simpler and missing a dumb feature than more complex with yagni baggage.

totten replied:
@colemanw Right, `$htmlize` has been pretty annoying-surprising. 

My hesitation for completely removing it:

* At some point, I expect we'd start wanting to convert existing calls (`CRM_Utils_System::url()` => `Civi::url()`). I bet 80% of those conversions are easier if you still have a flag for HTML.
* When encoding some data to go into an HTML document, there are different flavors (`htmlentities()`, `htmlspecialchars()`, `htmlspecialchars(...,ENT_QUOTES)`). One of those three is more susceptible to bugs (*re: CSRF, i18n, etc*) than the others. How often will people correctly remember which one?
* Just general gut sense that it's deeply ingrained, and that we'll bump into weird edges in removing it.

There's a sort of mirror-image between the PHP and Smarty (`{crmURL...}`) generation of URLs:

* In PHP, you should probably get the canonical URL most of the time -- but sometimes you want the HTML-encoded variant.
* In Smarty, you should probably get the HTML-encoded URL most of the time -- but sometimes you might another format (e.g. for JS-in-Smarty).

So at high level,  PHP and Smarty need different defaults but also need ways to flip over. It's kind of appealing that you have same list of allowed `$flags` regardless of the medium. Though TBH, I haven't really thought about the Smarty POV.

colemanw replied:
> I bet 80% of those conversions are easier if you still have a flag for HTML

Well... In a proper MVC, the escaping is done by the template layer. Definitely the smarty implementation ought to run the output of this function through `htmlentities()`, but it doesn't need a flag on this function to do that, it can just do it.
Angular pages definitely *don't* need the escaping, and neither do internal calls to something like `pushUserContext` or `redirect`. I can see that a lot of calls to `CRM_Utils_System::url()` accidentally leave escaping on when they shouldn't; for example (not to pick on anyone on purpose this is just the first one I found) @eileen just added one in 2023 that does escaping when it shouldn't:

https://github.com/civicrm/civicrm-core/blob/8ec2d293af07a1f2b65173a86a9afcd2488d32c1/CRM/Core/Form.php#L2711

I'm sure I've made that same mistake a dozen times or more. It's such a prevalent problem that we even have this aptly-named function:

https://github.com/civicrm/civicrm-core/blob/86d51faf4cfa430b2178b20c0c80fcbeafa41efc/CRM/Utils/String.php#L777-L790

So even if you're right and 80% of calls to `CRM_Utils_System::url` use escaping, I would hazard that probably 80% of those, *shouldn't*.

So what to do? Either we leave the flag in with a "this is dumb" warning, or take it out. I can live with it either way, as long as we don't ever attempt an auto-conversion of `CRM_Utils_System::url(...)` => `Civi::url()` that blindly copies over that flag. Every usage of `CRM_Utils_System` with escaping *on* is smelly and needs to be audited IMO.

colemanw replied:
While I was poking in there I noticed something really nutty: https://github.com/civicrm/civicrm-core/pull/26888

totten committed:
Civi::url() - Allow environments to change default scheme and format

totten committed:
Civi::url() - Add support for vars

totten committed:
(NFC) Civi::url() - Various docblocks

totten committed:
Civi::url() - Rename (p) to (t)

totten committed:
Civi::url() - Add support for path-variables

totten committed:
Civi::url() - Add option for resCacheCode

totten committed:
Civi::url() - Add test-coverage

totten committed:
Civi::url() - Split $fragment and $fragmentQuery. Define clear/consistent rule for escaping inputs.

totten committed:
Civi\Core\Url - Also support JSON serialization

totten committed:
UrlTest - Add coverage for variable-substitution

totten committed:
Civi\Core\Url - Allow dynamic variables

totten committed:
Smarty - Add support for `{url}frontend://civicrm/profile{/url}` (etal)

totten committed:
Civi::url() - Add support for `default://` URIs

totten commented:
@colemanw So I've done a bunch of updates

* Add test-coverage and many more docblocks.
* Add support for Angular-style URL data (`backend://civicrm/a/#/mailing/[id]?angularDebug=1`)
* Add better support for static asset URLs (eg `asset://[civicrm.packages]/foobar.js`)
* Add support for metadata-based routing (`default://civicrm/admin`)
* Define a consistent policy on escaping:
    * Whenever you pass in a string (`setQuery('id=123&msg=hello+world')`), then treat it as a literal URL fragment.
    * Whenever you pass an array (`setQuery(['id' => 123, 'msg' => 'hello world'])`), then we apply escaping.
* Define a Smarty utility `{url}...{/url}`:

    ```html
    <a href="{url}backend://civicrm/admin?reset=1{/url}">Administration</a>

    {url assign=adminUrl}backend://civicrm/admin?reset=1{/url}
    <a href="{$adminUrl}">Administration</a>

    <a href="{url contact=100 profile=200}current://civicrm/profile/view?id=[contact]&gid=[profile]{/url}">

    <a href="{url 1=$contacts[$i].contact_id 2=$profile.id}current://civicrm/profile/view?id=[1]&gid=[2]{/url}">
    ```

So overall, from my POV, it has all the functionality I've expected.

There are a couple possible areas of further consideration:

* [The Second Hard Problem of Computer Science](https://martinfowler.com/bliki/TwoHardThings.html) - The scheme names (`frontend://`, `backend://`, `asset://`, `assetBuilder://`) have a sort of logic to them. But are they good names?... `frontend://` and `backend://` are kinda long. Maybe `asset://` should be like `resource://` or `res://`. Yaddayadda.
* __Trial Conversions__ - Do a batch of versions (`CRM_Utils_System::url()` to `Civi::url()`; and `{crmURL}` to `{url}`). See if they feel alright.
* __URL Contexts/Default Options__ - So there's this recurring bug where you make a URL, but sometimes it doesn't look the way you expect, and so you have to go back and finesse options. 
    * Basically, when running code, you're in some operative context. And I think we come with different intuitions about how URLs should be formulated in each context.
        * __Displaying a web-page__. Here, relative URLs are nice. In case of ambiguity, you should lean toward keeping links in the same UI (i.e. frontend-pages tend to link to frontend-pages).
        * __Generating an email message__. Here, absolute URLs are nice. In case of ambiguity, you should lean toward frontend UI (even though the email might be generated from backend UI, from automated job, etc)
        * __Running admin/dev tools__ (`cv url`, `cv http`). Here, absolute URLs are nice. You're not in `frontend://` or `backend://`, so you could lean on metadata instead (`default://`).
    * I added some stuff to support these intuitions. [This seems to work for cv](https://github.com/civicrm/cv/pull/171). But the email-generating scenario seems important, and I haven't tried it. The concept feels tentative until the email side has been done. 

Those are all kind of debatable. Anyway, I'm not inclined to self-block any more... so I've removed "Draft/WIP" flag.

christianwach commented on Civi/Core/Url.php:
```diff
@@ -0,0 +1,690 @@
+
+    // TODO decide if the current default is good enough for future
+    $ssl = $this->getSsl() ?: \CRM_Utils_System::isSSL();
+    if ($ssl && str_starts_with($result, 'http:')) {
```
We'll need a shim for `str_starts_with()` otherwise CiviCRM will require PHP 8.0+.

totten replied:
@christianwach Yup, that's correct. [This shim](https://github.com/civicrm/civicrm-core/blob/5.63/composer.json#L100-L104) should take care of it, right?  (specifically https://github.com/symfony/polyfill-php80).

christianwach replied:
Ah cool, missed that.

christianwach reviewed (changes requested):
(Sorry, still not clear how the review process works here)

`str_starts_with()` is PHP 8.0+ so we'll need a shim to continue support for PHP 7.x.

christianwach commented on Civi/Core/Url.php:
```diff
@@ -0,0 +1,690 @@
+
+    switch ($scheme) {
+      case 'frontend':
+      case 'service':
+        $result = $userSystem->url($this->getPath(), $this->getQuery(), $preferFormat === 'absolute', $this->composeFragment(), TRUE, FALSE, FALSE);
+        break;
+
+      case 'backend':
+        $result = $userSystem->url($this->getPath(), $this->getQuery(), $preferFormat === 'absolute', $this->composeFragment(), FALSE, TRUE, FALSE);
+        break;
```
Am i right in thinking that routing to `$userSystem->url()` here provides an opportunity (sometime in the future) to replace it with something less problematic?

totten replied:
@christianwach I do like the idea of `$userSystem` having some more influence/visibility on the $scheme. For example, I don't think it's great for `frontend://` and `service://` to be conjoined like that. I'll add on some ideas for that...

totten replied:
@christianwach So does 59af62f8b3d280cf3080d78be3d9a86c28ab2749 seem better/worse/same? It delegates to `$userSystem->getRouteUrl()` instead of `$userSystem->url()`. The main differences are:

* `UF::getRouteUrl()` receives the name (`frontend`, `backend`, `service`) instead of multiple bools. It can potentially receive other/custom schemes if the UF has other environments.
* `UF::getRouteUrl()` doesn't have to deal with the fragment or absolute/relative flag. (These seem amenable to generic handling.)

Another thing the UF might influence is the interpretation of `current://` (ie `detectScheme()`). With the current PR, the UF-integration could [set `civicrm_url_defaults`](https://github.com/civicrm/cv/pull/171/files)  before invoking the page.

christianwach replied:
> So does https://github.com/civicrm/civicrm-core/commit/59af62f8b3d280cf3080d78be3d9a86c28ab2749 seem better/worse/same?

@totten Better in that the inscrutable boolean params are somewhat hidden. Nice one :)

christianwach commented:
@totten Thanks for starting to address this issue. I like the direction of travel - it's certainly a more readable way of collecting the various pieces of the URL together. Looks promising.

Coming from a WordPress-integration perspective, my focus is less on how the pieces are collected and more on how those URLs are subsequently built for use in the various CMS contexts, i.e.

1. WordPress admin
2. The Base Page with or without Clean URLs
3. Shortcodes - where URLs are based on the Permalink structure plus query vars
4. Blocks - which in [this plugin](https://develop.tadpole.cc/plugins/civicrm-gutenberg-blocks) call the QuickForm classes directly but may use `asset://` or `assetBuilder://` in future
5. WordPress multilingual plugins (which use the URL scheme in various ways to encode the desired language into it)

For now, it looks as though I can still work on the logic in `CRM_Utils_System_WordPress::url()` to make it adaptable to the various scenarios that it has to handle... just leaving this here as a reminder in case work is started on URL rendering that does not route via the current `$userSystem->url()` method(s).

totten committed:
Civi::url() - Simplify/normalize "fragment" and "relative" handling. Fix some edge-case bugs.

totten committed:
Extract CRM_Utils_System_Base::getRouteUrl()

totten committed:
Civi::url() - Allow building new URL from empty. Update comments.

totten commented:
Thanks for that last comment @christianwach. Those are some interesting contexts. Thinking about `#2` vs `#3` -- both are front-end-ish, but `#2` is a singular frontend -- and `#3` has n-many variations. Which means:

* While rendering web UI for `#2` or `#3`, you could use `current://` and expect links with matching structure.
* While processing a background job (*say, a sending a stock email with a link to update a recurring contribution*), it's much easier to link to `#2` than `#3`.  For `#2`, you can simply interpret `frontend://civicrm/contribute/updaterecur` as `https://example.com/civicrm/contribute/updaterecur`. But if you want the emails to point to one of n-many WP pages that support shortcodes, then you need to edit the email-template (or define some complicated override).

colemanw commented:
I think we've sat on this one long enough and it has good test coverage. Let's start using it!
